File: frontend/src/types/media.ts
export type MediaPlatform = "youtube";

export type MediaFormat = "mp4" | "mp3";

export type MediaQuality = "best" | `${number}p`;

export const SUPPORTED_MP3_BITRATES = [128, 192, 256, 320] as const;

export type AudioQuality = (typeof SUPPORTED_MP3_BITRATES)[number];

export interface AvailableMediaFormat {
  format_id: string;
  type: "video" | "audio";
  extension: string | null;
  quality: number | null;
  fps: number | null;
  bitrate: number | null;
  filesize: number | null;
}

export interface MediaInfo {
  id: string;
  title: string;
  author: string | null;
  duration: number | null;
  thumbnail: string | null;
  original_url: string;
  qualities: number[];
  formats: AvailableMediaFormat[];
}

export interface AnalyzeResponse {
  success: true;
  platform: MediaPlatform;
  media: MediaInfo;
}

export type DownloadRequest =
  | {
      url: string;
      format: "mp4";
      quality: number;
    }
  | {
      url: string;
      format: "mp3";
      audio_quality: AudioQuality;
    };

export interface DownloadResult {
  filename: string;
}

File: backend/app/schemas/media.py
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, HttpUrl


class AnalyzeRequest(BaseModel):
    url: HttpUrl = Field(description="Public YouTube URL to analyze")


class MP4DownloadRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    url: HttpUrl = Field(description="Public YouTube URL to download")
    format: Literal["mp4"]
    quality: int = Field(ge=1, le=4320, description="Exact video height in pixels")


class MP3DownloadRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    url: HttpUrl = Field(description="Public YouTube URL to download")
    format: Literal["mp3"]
    audio_quality: Literal[128, 192, 256, 320] = Field(
        description="Target MP3 bitrate in kbps"
    )


DownloadRequest = Annotated[
    MP4DownloadRequest | MP3DownloadRequest,
    Field(discriminator="format"),
]


class MediaFormat(BaseModel):
    format_id: str
    type: Literal["video", "audio"]
    extension: str | None = None
    quality: int | None = Field(default=None, ge=1)
    fps: float | None = Field(default=None, gt=0)
    bitrate: int | None = Field(default=None, ge=0)
    filesize: int | None = Field(default=None, ge=0)


class MediaInfo(BaseModel):
    id: str
    title: str
    author: str | None
    duration: int | None = Field(default=None, ge=0)
    thumbnail: HttpUrl | None
    original_url: HttpUrl
    qualities: list[int]
    formats: list[MediaFormat]


class AnalyzeResponse(BaseModel):
    success: Literal[True]
    platform: Literal["youtube"]
    media: MediaInfo
